[SPARK-50589][SQL] Avoid extra expensive expression duplication when push filter

zml1206 opened the pull request:
### What changes were proposed in this pull request?
First of all, when pushing down through Project and Aggregate, use With to rewrite non-cheap common expr instead of inline. This is equivalent to pushing down Filter and Project together.

Secondly, if the project is actually created in the end, there may be a risk of regression (some rules cannot match the filter using PhysicalOperation). so PushPredicateThroughProject is added after rewrite With, which uses the cost model to determine whether to push the predicate down through the project generated by With.


### Why are the changes needed?
1. Avoid nested expansion of filters expression causing execution plans to become too large. 
2. Avoid duplicate expensive expressions for better performance

### Does this PR introduce _any_ user-facing change?
No.


### How was this patch tested?
UT.


### Was this patch authored or co-authored using generative AI tooling?
No.

zml1206 commented:
@cloud-fan Can you help first to see if the overall design is feasible? It involves a lot of UT, so I will modify it step by step.

zml1206 commented:
> I'm very confused. When pushing predicates through Project, the problem we hit is we need to rewrite the attributes in the predicate with the corresponding expression in the Project. If the attribute appears more than once in the predicate, and the expression from Project is expensive, filter pushdown may make the query slower.
> 
> The solution is to wrap the predicate with `With` expression, adding the expression from Project as the common expression, and replace attributes in predicate with the common expression reference. But this PR does some rewriting and I don't understand what it does.

Filters support partial pushdown. If an attribute exists in two conditions, one condition is pushed to one level, and the other condition is pushed down to multiple levels. Therefore, after splitting by and, each filter need generates a With, then they need to share commonexprdef. Only the lowest commonexprdef will generate a common expression project. In addition, the child of the original common expression Alias ​​also needs to be replaced.

cloud-fan commented:
Why do we care about common expressions in the original predicate? This is an optimization opportunity but what we should focus on is to avoid perf regression when pushing filter through project that duplicate expensive expressions.

Maybe we should discuss this with a concrete example.

zml1206 commented:
> Why do we care about common expressions in the original predicate? This is an optimization opportunity but what we should focus on is to avoid perf regression when pushing filter through project that duplicate expensive expressions.
> 
> Maybe we should discuss this with a concrete example.

As the ut in pr, I hope that the common expression will only be calculated only once. `_common_expr_0` finally used in two Filters and `as e`.

zml1206 commented:
Do you want to make it simpler, commonexprdef is not shared, and the original alias is ignored?

cloud-fan commented:
OK let's make the scope clear. We do not aim to find all duplicated expressions in the query plan tree and optimize them with `With`. We aim to avoid extra expression duplication that happens during pushing filter through project. So the change should be we update the filter pushdown rule and use `With` to avoid the aforementioned expression duplication.

zml1206 commented:
> OK let's make the scope clear. We do not aim to find all duplicated expressions in the query plan tree and optimize them with `With`. We aim to avoid extra expression duplication that happens during pushing filter through project. So the change should be we update the filter pushdown rule and use `With` to avoid the aforementioned expression duplication.

OK, I'll do it as later, thank you.

zml1206 committed:
Avoid extra expression duplication when push filter

fix

fix

fix

fix

fix

update

update

fix

optimize split condition

optimize split condition

update

fix

fix

fix

fix

fix

fix

fix

update

fix

optimize Subquery

fix related InferFilters

rewrite With immediately

update

update

update

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/With.scala:
```diff
@@ -112,6 +112,22 @@ object With {
+   * @param replaceMap Replaced attributes and common expressions
+   */
+  def apply(expr: Expression, replaceMap: Map[Attribute, Expression]): With = {
+    val commonExprDefsMap = replaceMap.map(m => m._1 -> CommonExpressionDef(m._2))
```
why do we create a map if we never look up from it?

zml1206 replied:
Convenient to generate commonExprRefsMap.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1831,8 +1838,8 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
       }
 
       if (pushDown.nonEmpty) {
-        val pushDownPredicate = pushDown.reduce(And)
```
so the reason we don't simply rewrite the AND combined predicate is that we want to keep the predicate splittable later on?

zml1206 replied:
Yes

cloud-fan replied:
But then there is no common expression sharing between sub-predicates split by AND.

Shall we simply support splitting the predicate expression that is wrapped by `With`?

cloud-fan replied:
e.g. for `a * a > 0 AND a + a > 0` and `a` is a large expression, we first push down `With(large_expr, a * a > 0 AND a + a > 0)`, when we need to split it again later, we turn it into `With(large_expr, a * a > 0) AND With(large_expr, a + a > 0)`

cloud-fan replied:
OK I think this will go very complicated. We can't completely avoid duplicating large expressions because at some point, we can push down part of the predicates, so the common expression will appear in multiple plan nodes and we don't support cross-plan-nodes expression sharing yet.

zml1206 replied:
`select add,count(1) as ct from (select a+a as add,b from t) where add + add > 10 and add >ct`, 
the current implementation can push down add+add > 10 to aggregate.

cloud-fan replied:
More thoughts:
- I think it's common to reference the same column in multiple sub-predicates, e.g. `a > 0 AND a < 10`. The current approach doesn't help with this common case.
- We should improve the predicate split-and-combine process, to only accept duplicating large expressions when we have to, i.e. some sub-predicates need to be pushed down and some need to stay, and they share the same common expressions.

cloud-fan replied:
In the future, we should also support cross-plan-nodes expression sharing, by adding a Project to evaluate the common expression and propagate the result column all the way up.

zml1206 replied:
> We should improve the predicate split-and-combine process, to only accept duplicating large expressions when we have to, i.e. some sub-predicates need to be pushed down and some need to stay, and they share the same common expressions.

In this follow-up, we can merge the with expression before rewriteWith. This will not affect the push down.

zml1206 replied:
> In the future, we should also support cross-plan-nodes expression sharing, by adding a Project to evaluate the common expression and propagate the result attribute all the way up.


Yes, this is the ideal state. I planned to do this at first, so I made a lot of changes to with and encountered many problems.

cloud-fan replied:
> In this follow-up, we can merge the with expression before rewriteWith. This will not affect the push down.

It's better to not duplicate expressions in the first place. Creating a huge plan tree is harmful to the optimizer/planner.

zml1206 replied:
What I mean is that when cross-plan-nodes expression sharing is not supported, after the pushdown is completed and before rewriteWith, the with in the same filter can be merged.

zml1206 replied:
OK, It seems to be more troublesome, a + a> 5 and a < 10, a + a> 5 generates with, a < 10 is expanded.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/With.scala:
```diff
@@ -112,6 +112,22 @@ object With {
+   * @param expr original expression
+   * @param replaceMap Replaced attributes and common expressions
+   */
+  def apply(expr: Expression, replaceMap: Map[Attribute, Expression]): With = {
```
if the attribute only appears once in the expression, we can just inline, doesn't need to be a common expression.

zml1206 replied:
This logic in `PushPredicateThroughNonJoin.rewriteByWith()`.

zml1206 commented on sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q9/simplified.txt:
```diff
@@ -2,77 +2,72 @@ WholeStageCodegen (1)
-                        ColumnarToRow
-                          InputAdapter
-                            Scan parquet spark_catalog.default.store_sales [ss_quantity,ss_ext_discount_amt,ss_net_paid,ss_sold_date_sk]
+        HashAggregate [count,sum,count,sum,count] [count(1),avg(UnscaledValue(ss_ext_discount_amt)),avg(UnscaledValue(ss_net_paid)),mergedValue,count,sum,count,sum,count]
```
I don't understand why there is a change here. Is it because of CollapseProject? Is this correct? @cloud-fan

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala:
```diff
@@ -65,6 +65,8 @@ class SparkOptimizer(
       RewriteDistinctAggregates),
     Batch("Pushdown Filters from PartitionPruning", fixedPoint,
       PushDownPredicates),
+    Batch("Rewrite With expression", fixedPoint,
```
shall we just rewrite `With` once at the very end of the optimizer?

zml1206 replied:
Let me try.

zml1206 replied:
We can't push With filter into scan, so rewrite with at the end of `operatorOptimizationBatch`.

cloud-fan commented:
how about the rule `PushPredicateThroughNonJoin`?

zml1206 commented:
> how about the rule `PushPredicateThroughJoin`?

There is no replaceAlias ​​logic in PushPredicateThroughJoin.

cloud-fan commented:
ah ok. Shall we implement the better predicate split-and-combine in this PR to make it more useful? Otherwise we can't handle common cases like `a + a > 5 AND a < 10`.

zml1206 commented:
@cloud-fan It seems that the change in filters order causes the execution plan to mismatch. Is there any good solution?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:
```diff
@@ -107,6 +107,14 @@ trait PredicateHelper extends AliasHelper with Logging {
   }
 
+  /**
+   * First split the predicates by And, then combine the predicates with the same references.
```
I think there is a misunderstanding. What I proposed is `splitConjunctivePredicates` should split AND/OR under `With`: `With(c1, c2, c1 + c1 > 0 AND c2 + c2 > 0 AND c1 + c2 > 0)` can be split into `With(c1, c1 + c1 > 0)`, `With(c2, c2 + c2 > 0)`, ` With(c1, c2, c1 + c2 > 0)`. When combine the predicates later, we also recognize the `With` and merge them to avoid duplicated expressions.

zml1206 replied:
why split OR

zml1206 replied:
Merge `With` requires a certain cost. It seems that we only need to merge once before rewriting `With`.

cloud-fan replied:
note: when splitting the predicates under `With` (by `AND`), we don't refresh the CTE ids, so when merging the `With` back, it should be pretty easy.

zml1206 replied:
It is unnecessary to continuously split and merge with. We only need to split by and to generate with, and finally merge with before rewriting with.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/With.scala:
```diff
@@ -112,6 +112,22 @@ object With {
   }
 
+  /**
+   * Helper function to create a [[With]] statement when push down filter.
```
Since this is very specific to filter pushdown, shall we put this method in the filter pushdown rule?

zml1206 replied:
OK, wait for CI to finish running.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1831,8 +1841,8 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
       if (pushDown.nonEmpty) {
-        val pushDownPredicate = pushDown.reduce(And)
-        val replaced = replaceAlias(pushDownPredicate, aliasMap)
+        val replacedByWith = rewriteConditionByWith(pushDown.reduce(And), aliasMap)
```
`rewriteConditionByWith` will split the predicate anyway, why do we combine them with `And` here?

zml1206 replied:
Calculate the number of times each common expression is used through the entire condition.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1831,8 +1841,8 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
-        val pushDownPredicate = pushDown.reduce(And)
-        val replaced = replaceAlias(pushDownPredicate, aliasMap)
+        val replacedByWith = rewriteConditionByWith(pushDown.reduce(And), aliasMap)
+        val replaced = replaceAlias(replacedByWith, aliasMap)
```
doesn't `rewriteConditionByWith` already replace the aliases?

zml1206 replied:
No, `rewriteConditionByWith` only rewrite common attribute to common expression ref.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala:
```diff
@@ -96,14 +96,17 @@ object RewriteWithExpression extends Rule[LogicalPlan] {
         val defs = w.defs.map(rewriteWithExprAndInputPlans(_, inputPlans, isNestedWith = true))
         val refToExpr = mutable.HashMap.empty[CommonExpressionId, Expression]
         val childProjections = Array.fill(inputPlans.length)(mutable.ArrayBuffer.empty[Alias])
+        val refsCount = child.collect { case r: CommonExpressionRef => r}
```
we should only collect references to the current `With`, not the nested `With`.
```suggestion
        val refsCount = child.collect { case r: CommonExpressionRef if defs.exists(_.id == r.id) => r}
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeWithExpression.scala:
```diff
@@ -0,0 +1,69 @@
+    val newDefs = left.defs.toBuffer
+    val replaceMap = mutable.HashMap.empty[CommonExpressionId, CommonExpressionRef]
+    right.defs.foreach {rDef =>
+      val index = left.defs.indexWhere(lDef => rDef.child.fastEquals(lDef.child))
```
plan comparison can be expensive. Since it's a very special case for filter pushdown, shall we use the same `CommonExpressionDef` ids for different `With`s in the split predicates?

zml1206 commented:
@cloud-fan https://github.com/zml1206/spark/actions/runs/12478377527/job/34825933696, Do you know the reason for the pyspark ci failure?

zml1206 commented:
CI failure is not relevant.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1978,6 +1988,62 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
+    if (!SQLConf.get.getConf(SQLConf.ALWAYS_INLINE_COMMON_EXPR)) {
+      val canRewriteConf = cond.filter(canRewriteByWith)
+      if (canRewriteConf.nonEmpty) {
+        val replaceWithMap = canRewriteConf.reduce(And)
```
we can do `canRewriteConf.flatMap(_.collect...)`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1978,6 +1988,62 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
+
+  // With does not support inline subquery
+  private def canRewriteByWith(expr: Expression): Boolean = {
+    !expr.containsPattern(PLAN_EXPRESSION)
```
is this check too strong? We only require the common expression to not contain subqueries.

zml1206 replied:
It seems not feasible, https://github.com/zml1206/spark/actions/runs/12491712984/job/34858074548

zml1206 replied:
Both rewriting subqueries and pushing down predicates are in batch "operator optimization before inferring filters". Pushing down predicates may cause SubqueryExpression to contain common refs, and then rewriting subqueries cannot replace common refs.

cloud-fan replied:
so we only need to make sure `SubqueryExpression#reference` does not contains the to-be-replaced attributes?

zml1206 replied:
It seems ok, I'll give it a try, thank you.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1978,6 +1988,62 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
+          .groupBy(identity)
+          .transform((_, v) => v.size)
+          .filter(m => aliasMap.contains(m._1) && m._2 > 1)
+          .map(m => m._1 -> trimAliases(aliasMap.getOrElse(m._1, m._1)))
```
Slo we've done the work of `def replaceAlias` here, why do we need to call `replaceAlias`?

zml1206 replied:
Here just get the map of attribute and common expression to generate With, without rewriting the condition.

cloud-fan replied:
The common expression is already the actual expression in the `aliasMap`, and the corresponding attributes in the condition are rewritten with common expression ref, what else need to be rewritten?

zml1206 replied:
There are also common expressions that have not been overridden with ref, for example, cheap or once-used common expression

zml1206 replied:
Only the common expressions that need to be replaced with ref are rewritten here.

cloud-fan replied:
oh I see.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1978,6 +1988,62 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
+    val defs = mutable.HashSet.empty[CommonExpressionDef]
+    val replaced = expr.transform {
+      case a: Attribute if refsMap.contains(a) =>
+        defs.add(defsMap.get(a).get)
```
```suggestion
        defs.add(defsMap(a))
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala:
```diff
@@ -96,14 +96,22 @@ object RewriteWithExpression extends Rule[LogicalPlan] {
+        val refsCount = child.collect {
+            case r: CommonExpressionRef
+              if defs.exists {
+                case d: CommonExpressionDef => d.id == r.id
```
nit: we can create an id set to avoid repeated linear search.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala:
```diff
@@ -95,7 +100,11 @@ class SparkOptimizer(
-    Batch("Replace CTE with Repartition", Once, ReplaceCTERefWithRepartition)))
+    Batch("Replace CTE with Repartition", Once, ReplaceCTERefWithRepartition),
+    Batch("Merge With expression", fixedPoint, MergeWithExpression),
+    Batch("Rewrite With expression", fixedPoint,
```
how many places do we need to put this batch?

zml1206 replied:
Before Extract Python UDFs, infer filters and last of each optimizer.

cloud-fan replied:
Can you elaborate on the issue if don't rewrite `With` before Extract Python UDFs and infer filters?

zml1206 replied:
Extract Python UDFs will extracts PythonUDFs from operators  to `BatchEvalPython`/`ArrowEvalPython`, so it is necessary to ensure that EvalPython does not contain common expression ref.

zml1206 replied:
There seems to be no problem with infer filters. As long as ensure infer filters once, I will fix it.

cloud-fan replied:
OK now this makes me a bit worried. One major limitation of `With` is that we must move it as a whole. Otherwise we have dangling `CommonExpressionRef`.

It seems safer to rewrite `With` immediately at the beginning of the optimizer and right after the place that we produce new `With`.

zml1206 replied:
Yes, we must move `With` as a whole, so rewrite `With` every time after push down predicate?

cloud-fan replied:
Yea, for now we are sure that only two places can produce `With`: the analyzer and the predicate pushsdown rule. It's safer to take care of these two.

cloud-fan commented on sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q21.sf100/explain.txt:
```diff
@@ -131,38 +133,46 @@ Functions [2]: [sum(CASE WHEN (d_date#10 < 2000-03-11) THEN inv_quantity_on_hand
+(23) Project [codegen id : 5]
+Output [5]: [w_warehouse_name#12, i_item_id#7, inv_before#19, inv_after#20, CASE WHEN (inv_before#19 > 0) THEN (cast(inv_after#20 as double) / cast(inv_before#19 as double)) END AS _common_expr_0#21]
 Input [4]: [w_warehouse_name#12, i_item_id#7, inv_before#19, inv_after#20]
-Condition : (CASE WHEN (inv_before#19 > 0) THEN ((cast(inv_after#20 as double) / cast(inv_before#19 as double)) >= 0.666667) END AND CASE WHEN (inv_before#19 > 0) THEN ((cast(inv_after#20 as double) / cast(inv_before#19 as double)) <= 1.5) END)
```
ah so this is a positive plan change that we avoid computing the CASE WHEN multiple times in the Filter.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeWithExpression.scala:
```diff
@@ -0,0 +1,45 @@
+ * Before rewrite with expression, merge with expression which has same common expression for
+ * avoid extra expression duplication.
+ */
+object MergeWithExpression extends Rule[LogicalPlan] {
```
I think it's more efficient to just do it in the rewriting rule, see https://github.com/apache/spark/pull/49303

cloud-fan replied:
It's merged, can you rebase your PR?

zml1206 replied:
How about I first separate a pr to inline the common expression in With if used once?

cloud-fan replied:
sure

zml1206 replied:
#49310, Can you help take a look, thanks @cloud-fan

cloud-fan commented:
This LGTM overall, my last suggestion is to narrow the scope of the `With` expression generated during filter pushdown. We should rewrite the `With` expression immediately at the end of the filter pushdown rule. However, this causes conflicts as we need to push filters through the `Project` generated by `With` in the next iteration, and we need to have a way to make it idempotent.

My idea is, when rewriting `With` inside `Filter`, and the generated `Project` is the same as the `Project` above `Filter`, we remove the above `Project`.

zml1206 commented:
> My idea is, when rewriting `With` inside `Filter`, and the generated `Project` is the same as the `Project` above `Filter`, we remove the above `Project`.

There is a problem in this, it cannot push down `Filter` to the child of the `Project`.

zml1206 commented:
> My idea is, when rewriting `With` inside `Filter`, and the generated `Project` is the same as the `Project` above `Filter`, we remove the above `Project`.

There is a problem. The new iteration generates different refs. Do you have any suggestions?

cloud-fan commented:
After more thought, I think we should think of filter pushdown in a different way. Once you push a predicate through `Project` and expand the attribute reference into an expensive expression, even only once, there is a risk of perf regression, because that expensive expression will be evaluated twice: once in the `Filter` being pushed down, once in the `Project` stays up.

A safe approach is that when pushing filters through `Project`, the filters should promise to produce the expensive expressions as attributes, so that we can rewrite the `Project` during the pushdown to use these attributes, to make sure the expensive expressions are only evaluated once. We can extend the `With` expression to use pre-defined Alias to support it.

zml1206 commented:
> After more thought, I think we should think of filter pushdown in a different way. Once you push a predicate through `Project` and expand the attribute reference into an expensive expression, even only once, there is a risk of perf regression, because that expensive expression will be evaluated twice: once in the `Filter` being pushed down, once in the `Project` stays up.
> 
> A safe approach is that when pushing filters through `Project`, the filters should promise to produce the expensive expressions as attributes, so that we can rewrite the `Project` during the pushdown to use these attributes, to make sure the expensive expressions are only evaluated once. This is kind of we pushing down filters and part of the `Project` together. We can extend the `With` expression to use pre-defined Alias to support it.


This seems to be back to sharing across nodes `With`.

zml1206 committed:
rewrite With expression immediately at the end of the filter pushdown rule

zml1206 committed:
project alias reuse attribute

zml1206 committed:
CollapseProject once

zml1206 committed:
fix

zml1206 committed:
fix ut

zml1206 commented:
cc @cloud-fan

zml1206 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala:
```diff
@@ -38,8 +38,34 @@ import org.apache.spark.util.Utils
+  // once in the Project stays up. Project reuses the attribute to avoid secondary evaluates.
+  private def rewriteAlias(plan: LogicalPlan): LogicalPlan = {
+    plan.transformUpWithSubqueriesAndPruning(_.containsAllPatterns(PROJECT, FILTER)) {
+      case p @ Project(projectList, child: Filter)
```
The aggregateExpressions of aggregate are not rewritten here, because the groupingExpressions will be changed, which may cause regression.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -104,6 +104,7 @@ abstract class Optimizer(catalogManager: CatalogManager)
         ReorderJoin,
         EliminateOuterJoin,
         PushDownPredicates,
+        RewriteWithExpression,
```
it's fragile to rely on rule order, let's call `RewriteWithExpression` directly at the end of `PushDownPredicates`

zml1206 committed:
call RewriteWithExpression directly at the end of PushDownPredicates

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1978,6 +1977,55 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
+      val replaceWithMap = cond.collect {case a: Attribute => a }
+        .groupBy(identity)
+        .transform((_, v) => v.size)
+        .filter(m => aliasMap.contains(m._1) && m._2 > 1)
```
I don't think we need to check ref count anymore. Once we push down predicate through a project/aggregate, we already introduce repeated execution (Filter and Project/Aggregate) and we should use With to optimize it.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1811,7 +1811,7 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
       if fields.forall(_.deterministic) && canPushThroughCondition(grandChild, condition) =>
       val aliasMap = getAliasMap(project)
-      project.copy(child = Filter(replaceAlias(condition, aliasMap), grandChild))
+      project.copy(child = Filter(rewriteCondition(condition, aliasMap), grandChild))
```
I think we should rewrite the `Project` here. A predicate with `With` expression means an associated `Project` to output the common expressions. We should rewrite the project expressions that are referenced by the predicate into `Attribute`.

zml1206 replied:
If the filter is pushed down multiple nodes, the attribute will be lost.

cloud-fan replied:
This also means that when we push down predicate through other operators, they should propagate the attributes. Looking at `def canPushThrough`, I think most of them are fine (they propagate all output attribute from child), but `Expand` need update, we need to append the attributes that will be generated by `With` to `Expand#projections` and `output`.

zml1206 replied:
Aggregation is also need update, but groupingExpressions will be changed, which may cause regression.

cloud-fan replied:
can you elaborate about the possible regression?

zml1206 replied:
Introduce complex types or variable length types group keys, before is HashAgg, after is SortAgg

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1783,12 +1783,12 @@ object PruneFilters extends Rule[LogicalPlan] with PredicateHelper {
-      .orElse(PushPredicateThroughNonJoin.applyLocally)
-      .orElse(PushPredicateThroughJoin.applyLocally)
-  }
+  def apply(plan: LogicalPlan): LogicalPlan = RewriteWithExpression(
```
We can create a dedicated RuleExecutor to run this rule
```
object PushDownPredicates extends Rule[LogicalPlan] {
  val rewriteWithRuleExecutor = new RuleExecutor[LogicalPlan] {
    val batches = Batch("RewriteWithExpression", fixedPoint, RewriteWithExpression) :: Nil
  }
}
```

zml1206 committed:
refactor

zml1206 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala:
```diff
@@ -90,7 +90,17 @@ object RewriteWithExpression extends Rule[LogicalPlan] {
     // child, and we need to project away the extra columns to keep the plan schema unchanged.
     assert(p.output.length <= newPlan.output.length)
-    if (p.output.length < newPlan.output.length) {
+    newPlan.output.diff(p.output).forall(_.name.startsWith("_"))
```
The redundant ones will be deleted later.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1132,6 +1132,7 @@ object CollapseProject extends Rule[LogicalPlan] with AliasHelper {
         r.copy(child = p.copy(projectList = buildCleanedProjectList(l1, p.projectList)))
       case Project(l1, s @ Sample(_, _, _, _, p2 @ Project(l2, _))) if isRenaming(l1, l2) =>
         s.copy(child = p2.copy(projectList = buildCleanedProjectList(l1, p2.projectList)))
+      case p @ Project(_, child) if child.output == p.output => child
```
Is it the same optimization in `RemoveNoopOperators`?

zml1206 replied:
Yea, thank you.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1978,6 +2018,70 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
+      cond: Expression,
+      aliasMap: AttributeMap[Alias]): Expression = {
+    if (!SQLConf.get.getConf(SQLConf.ALWAYS_INLINE_COMMON_EXPR)) {
+      val replaceWithMap = cond.collect {case a: Attribute => a }
```
nit: `cond.references.toSeq.filter...`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1810,8 +1825,17 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
-      project.copy(child = Filter(replaceAlias(condition, aliasMap), grandChild))
+      var newProjectList = fields ++ getWithAttributes(condition)
+      val newCondition = rewriteCondition(condition, aliasMap)
+      val rewriteAlias = getWithAlias(newCondition).toSet
```
We should just return a set of ExprId, and check `idSet.contains(alias.exprId)`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1831,9 +1855,15 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
+        // So Aggregate only need inline common expression from parent for original project
+        // inheritance.
+        val newAggregateExpressions =
+          aggregate.aggregateExpressions ++ getWithAlias(pushDown.reduce(And))
```
```suggestion
          aggregate.aggregateExpressions ++ replaceAlias(getWithAlias(pushDown.reduce(And)))
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1978,6 +2018,70 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
+          m._1 -> CommonExpressionDef(child = trimAliases(m._2), originAlias = Some(m._2))))
+        val refsMap = AttributeMap(defsMap.map(m => m._1 -> new CommonExpressionRef(m._2)))
+        splitConjunctivePredicates(cond)
+          .map(rewriteByWith(_, defsMap, refsMap))
```
shall we wrap the entire condition with a single `With`? This is to make sure we don't duplicate a huge expression tree in the query plan.

We can temporarily strip `With` when splitting the condition when pushing down predicate through Aggregate, and then duplicate the `With` if we have to.

zml1206 replied:
Splitting With is tricky. Each pushdown requires splitting and merging, and most support partial pushdowns, not just aggregate.

zml1206 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/With.scala:
```diff
@@ -149,8 +149,14 @@ object CommonExpressionId {
+case class CommonExpressionDef(
+    child: Expression,
+    id: CommonExpressionId = new CommonExpressionId(),
+    originAlias: Option[Alias] = None)
```
It seems not possible. The origin project needs to inherit this attribute, and the exprId is not enough.

zml1206 replied:
It seems more appropriate to use attribute.

zml1206 committed:
optimize partial pushdown: extend attribute when child already contains it

zml1206 committed:
fix

zml1206 committed:
fix

zml1206 committed:
fix style

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1281,8 +1281,8 @@ object CollapseProject extends Rule[LogicalPlan] with AliasHelper {
       }
-    // Alias and ExtractValue are very cheap.
-    case _: Alias | _: ExtractValue => e.children.forall(isCheap)
+    // Alias, ExtractValue and CreateNamedStruct are very cheap.
```
create struct is cheap?

zml1206 committed:
fix some test

zml1206 committed:
add cost model

zml1206 committed:
fix some ut

zml1206 committed:
fix

zml1206 committed:
fix

zml1206 committed:
fix

zml1206 committed:
update comments
